datalad/core/local/diff.py

```diff
@@ -0,0 +1,335 @@
+    These state identifiers are evaluated in the context of the (specified
+    or detected) dataset. In case of a recursive report on a dataset
+    hierarchy corresponding state pairs for any subdataset are determined
+    from the subdataset record in the respective superdataset. Only changes
+    recorded in a subdataset between these two states are reported, and so on.
+
+    Any paths given as additional arguments will be used to constrain the
+    difference report. As with Git's diff, it will not result in an error when
+    a path is specified that does not exist on the filesystem.
+
+    Reports are very similar to those of the `rev-status` command, with the
+    distinguished content types and states being identical.
+    """
+    # make the custom renderer the default one, as the global default renderer
+    # does not yield meaningful output for this command
+    result_renderer = 'tailored'
+
+    _params_ = dict(
+        _common_diffstatus_params,
+        path=Parameter(
+            args=("path",),
+            metavar="PATH",
+            doc="""path to contrain the report to""",
+            nargs="*",
+            constraints=EnsureStr() | EnsureNone()),
+        fr=Parameter(
+            args=("-f", "--from",),
+            dest='fr',
+            metavar="REVISION",
+            doc="""original state to compare to, as given by any identifier
+            that Git understands.""",
+            nargs=1,
+            constraints=EnsureStr()),
+        to=Parameter(
+            args=("-t", "--to",),
+            metavar="REVISION",
+            doc="""state to compare against the original state, as given by
+            any identifier that Git understands. If none is specified,
+            the state of the worktree will be used compared.""",
+            nargs=1,
+            constraints=EnsureStr() | EnsureNone()),
+    )
+
+    @staticmethod
+    @datasetmethod(name='diff')
+    @eval_results
+    def __call__(
+            path=None,
+            fr='HEAD',
+            to=None,
+            dataset=None,
+            annex=None,
+            untracked='normal',
+            recursive=False,
+            recursion_limit=None):
+        ds = require_dataset(
+            dataset, check_installed=True, purpose='difference reporting')
+
+        # convert cmdline args into plain labels
+        if isinstance(fr, list):
+            fr = fr[0]
+        if isinstance(to, list):
+            to = to[0]
+
+        for r in _diff_cmd(
+                ds=ds,
+                dataset=dataset,
+                fr=fr,
+                to=to,
+                constant_refs=False,
+                path=path,
+                annex=annex,
+                untracked=untracked,
+                recursive=recursive,
+                recursion_limit=recursion_limit):
+            yield r
+
+    @staticmethod
+    def custom_result_renderer(res, **kwargs):  # pragma: no cover
```

Comment:
I dislike those `pragma: no cover` for pieces of code which can and should be tested. If we do not test them, that is "ok" but I think we shouldn't pretend that they shouldn't be

Reply:
I recently started doing that. The reason is that I can see right in the code what does not have a dedicated test (i.e. a test that is intentionally written for a code piece). I find that coverage has fooled me frequently into thinking things are being tested. I dont see it as pretending, rather than a declaration of test absence. I cannot come up with a reason why any piece of code should not be tested.

Reply:
Just visit codecov and you will see all lines which aren't covered by tests. Such pragma comments is intended to be used for code which isn't/can't be tested - those are ignored while computing % of code covered

Reply:
FTR: I know that I can get lines that are not executed at all from coverage reports. What I described is something different. I want to mark a code block that I know is not tested systematically (it may be executed, hence "covered", but not under all relevant circumstances). An increasingly relevant issue for me are list comprehensions (that I started to use much more frequently). They are generally full of conditionals, but regardless of whether I test them, the list comprehension is always covered completely. For the case that you commented on that distinction is not relevant, because as of now that code is not executed by the tests (I think). But I don't care much if it is. The point is that even if executed, it is not tested.

Reply:
ah, thanks for the explanation of the purpose. So we should them come up with some additional/different annotation for those to not confuse with "exclude from what needs to be covered" which is what `pragma: no cover` is used for.  May be `pragma: more cover` or alike?

<details>
<summary>BTW, just discovered to myself ["branch coverage"](https://coverage.readthedocs.io/en/coverage-4.3.4/branch.html#branch) in coverage.py.</summary>

```shell
$> python-coverage run cov.py         
(dev) 1 10091.....................................:Fri 03 May 2019 11:32:29 AM EDT:.
hopa:/tmp
$> python-coverage report    
Name     Stmts   Miss  Cover
----------------------------
cov.py       8      0   100%
(dev) 1 10092.....................................:Fri 03 May 2019 11:32:31 AM EDT:.
hopa:/tmp
$> python-coverage run --branch cov.py
(dev) 1 10093.....................................:Fri 03 May 2019 11:32:38 AM EDT:.
hopa:/tmp
$> python-coverage report             
Name     Stmts   Miss Branch BrPart  Cover
------------------------------------------
cov.py       8      0      2      1    90%
(dev) 1 10094.....................................:Fri 03 May 2019 11:32:40 AM EDT:.
hopa:/tmp
$> python-coverage annotate           
(dev) 1 10095.....................................:Fri 03 May 2019 11:32:55 AM EDT:.
hopa:/tmp
$> cat cov.py,cover        
> def my_partial_fn(x):       # line 1
>     if x:                   #      2
>         y = 10              #      3
>     return y                #      4
  
> y = 1
> my_partial_fn(y)
  # list comprehension where portion is never triggered as well
> [i for i in range(2, 10) if i > y]
  # or assignment
> z = y or None
```
</details>

Reply:
https://github.com/datalad/datalad/issues/3382 to continue on this topic